pkg/collector/container.go

```diff
@@ -129,7 +143,108 @@ func getContainerPorts(logger logr.Logger, cfg string) map[string]corev1.Contain
 	return ports
 }
 
-func portMapToList(portMap map[string]corev1.ContainerPort) []corev1.ContainerPort {
+func getServicePortsFromCWAgentConfig(logger logr.Logger, configJSON *adapters.CwaConfig) []corev1.ServicePort {
+	servicePortsMap := getAppSignalsServicePortsMap()
+	getMetricsReceiversServicePorts(logger, configJSON, servicePortsMap)
+	getLogsReceiversServicePorts(logger, configJSON, servicePortsMap)
+	getTracesReceiversServicePorts(logger, configJSON, servicePortsMap)
+	return PortMapToServicePortList(servicePortsMap)
+}
+
+func getMetricsReceiversServicePorts(logger logr.Logger, configJSON *adapters.CwaConfig, containerPortsMap map[int32]corev1.ServicePort) {
+	if configJSON.Metrics == nil || configJSON.Metrics.MetricsCollected == nil {
+		return
+	}
+	//StatD - https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-custom-metrics-statsd.html
+	if configJSON.Metrics.MetricsCollected.StatsD != nil {
+		getReceiverServicePort(logger, configJSON.Metrics.MetricsCollected.StatsD.ServiceAddress, StatsD, corev1.ProtocolUDP, containerPortsMap)
+	}
+	//CollectD - https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-custom-metrics-collectd.html
+	if configJSON.Metrics.MetricsCollected.CollectD != nil {
+		getReceiverServicePort(logger, configJSON.Metrics.MetricsCollected.CollectD.ServiceAddress, CollectD, corev1.ProtocolUDP, containerPortsMap)
+	}
+}
+
+func getReceiverServicePort(logger logr.Logger, serviceAddress string, receiverName string, protocol corev1.Protocol, containerPortsMap map[int32]corev1.ServicePort) {
+	if serviceAddress != "" {
+		port, err := portFromEndpoint(serviceAddress)
+		if err != nil {
+			logger.Error(err, "error parsing port from endpoint")
```

Comment:
if there is an error parsing the endpoint, should we fall back to using the default?

Reply:
I think yes ! I was also thinking what should happen if the customer supplies an invalid config (ex - statsd config without the service address field), should the operator open default ports ? We should mimic the CWA behavior in that case

Reply:
If they don't mention the service address at all in the config, code already handles to add the default address. But if the service address given in the config is not valid, I think it doesn't make sense to make it default as customer might not be using that port at all.